### ISSUE TITLE
Uneven pixels in native resolution with integer scaling and without filtering

### ISSUE THREAD

**Issue by levonnn** [stenzek/duckstation#1564]
**Game details**
SLUS-00067 Castlevania: Symphony of the Night (USA)

**Description of the issue / bug**
Uneven pixels in native resolution with integer scaling and without filtering. Some of them are wider than other. Even native resolution without upscale displays chewed-up pixels.

**Steps to reproduce**
1) Fresh unpack of the latest Win 64 build. No changes to settings except for:
Linear filtering - off
Integer scaling - on
Aspect ration - 4:3
2) Launch the game and reach the main menu. But it's visible anywhere in the game.

**Enhancements information**
Enhancements are turned off.

**Hardware/software**
i7 920, GTX 760, Win 7 64.
GPU Renderer - D3D11

**Emulator version**
0.1-3042

**Additional context**
How it looks without ehancements:
![Untitled](https://user-images.githubusercontent.com/46363559/106574388-e8588500-654b-11eb-8d97-7d2ad8120edb.png)

How it looks with 5x upscale and linear filtering on. The differences is especially visible on "X" button and dithered pattern.
![Untitled1](https://user-images.githubusercontent.com/46363559/106574397-ea224880-654b-11eb-9713-956231ba7477.png)

**Comment by mitigd** [comment]
From what I can tell this happens when setting the option 'crop' to 'only overscan area' in the 'Screen Display' area in 'Display Settings'. Not exactly sure if that makes things _correct_ but it does get rid of that distortion on the x button for me.

**Comment by Farlander-001** [comment]
This game is rare exception, running at 256*224 pixels resolution. As a workaround, I set internal resolution as 3x, and enabled linear filtering. Result - sharp pixels vertically and light filtering horizontally. Sorry for poor English)

**Comment by levonnn** [comment]
> This game is rare exception, running at 256*224 pixels resolution. As a workaround, I set internal resolution as 3x, and enabled linear filtering. Result - sharp pixels vertically and light filtering horizontally. Sorry for poor English)

In other words, if we don't want to upres 3D elements by playing in native resolution, we need sharp billinear or a similar filter to scale things properly.

**Comment by stenzek** [comment]
I'm still a little confused here. So the game is displaying 365x224, which is wider than 4:3. DuckStation applies the 4:3 display aspect ratio by scaling the width, and with a non-integer scaling on the width (~0.8 or so), you will get the odd duplicated pixel, as the resulting pixels are not square. This would be the same as on console.

Perhaps I'm missing where the issue is.

Also, there's a "PAR 1:1" aspect ratio option, which will give you square pixels. But technically not the correct AR for the game as it would be on a TV in this case.

**Comment by Farlander-001** [comment]
> I'm still a little confused here. So the game is displaying 365x224, which is wider than 4:3. DuckStation applies the 4:3 display aspect ratio by scaling the width, and with a non-integer scaling on the width (~0.8 or so), you will get the odd duplicated pixel, as the resulting pixels are not square. This would be the same as on console.
> 
> Perhaps I'm missing where the issue is.
> 
> Also, there's a "PAR 1:1" aspect ratio option, which will give you square pixels. But technically not the correct AR for the game as it would be on a TV in this case.

In game resolution changes to 256*224. In menus - 365*224

**Comment by Farlander-001** [comment]
Having ability to apply only horizontal or only vertical filtering can be a good solution

**Comment by levonnn** [comment]
> I'm still a little confused here. So the game is displaying 365x224, which is wider than 4:3. DuckStation applies the 4:3 display aspect ratio by scaling the width, and with a non-integer scaling on the width (~0.8 or so), you will get the odd duplicated pixel, as the resulting pixels are not square. This would be the same as on console.
> 
> Perhaps I'm missing where the issue is.
> 
> Also, there's a "PAR 1:1" aspect ratio option, which will give you square pixels. But technically not the correct AR for the game as it would be on a TV in this case.

The original point of discussion was that I said that I'd happily migrate to DuckStation as PS1 emulator of choice if there was an option to even pixels (not square, but of not varying proportions) via filters while retaining relatively crisp pixels at native resolution. It was actually requested here some time ago as an enhancement. Some guy on reddit argued that if you use integer scaling, pixels won't be distorted.

In other words, as it was requested, we need sharp bilinear or anything similar for stand-alone DuckStation.

**Comment by stenzek** [comment]
Maybe I'm still missing the point, but if you're filtering the pixels, they're not going to be perfectly consistent proportion-wise either? (think of stretching one grid over another - there will be some cells which don't perfectly overlap).

**Comment by ggrtk** [comment]
> Maybe I'm still missing the point, but if you're filtering the pixels, they're not going to be perfectly consistent proportion-wise either? (think of stretching one grid over another - there will be some cells which don't perfectly overlap).

I'd also agree that this would be the case. I'm not familiar with sharp bilinear, so I'd guess it's probably doing something like only applying linear filtering for host display pixels that are crossing texel boundaries, which still does come at a cost of some sharpness.

For this particular case, the only real solution that I'm aware of for getting a sharp and consistently-sized pixel look on a uniform display grid at the original aspect ratio is to combine point sampling with an extremely high display resolution.

**Comment by levonnn** [comment]
> Maybe I'm still missing the point, but if you're filtering the pixels, they're not going to be perfectly consistent proportion-wise either? (think of stretching one grid over another - there will be some cells which don't perfectly overlap).

Filtering won't be needed on a big screen, because you can safely use integer scale up to something like 5-6x. But on something like 1080p or less filtering is a compromise. It makes pixel grid more uniform with slight loss of crispness. Like on the second example in the OP.

**Comment by stenzek** [comment]
Do you have any references or examples of this "sharp bilinear filtering"? I haven't heard of it before.

**Comment by dankcushions** [comment]
there's some good examples of a good sharp pixel filter here: https://github.com/gonetz/GLideN64/issues/2116
which sources from https://colececil.io/blog/2017/scaling-pixel-art-without-destroying-it/

btw, such filters can hammer low powered devices, so should always be optional.

**Comment by stenzek** [comment]
Sounds like a good candidate to use as a "scaling shader" option then. I actually had those mostly implemented, will have to dig up the branch.

**Comment by levonnn** [comment]
> Do you have any references or examples of this "sharp bilinear filtering"? I haven't heard of it before.

bsnes comes with shaders like "AANN" and "Sharp-Bilinear", for example.

**Comment by Konkonfaux** [comment]
> Do you have any references or examples of this "sharp bilinear filtering"? I haven't heard of it before.

It is under interpolation shaders in  Retroarch. I think you can see what it looks like on Duckstation using the Swanstation core.

**Comment by Lolibox** [comment]
https://github.com/rsn8887/Sharp-Bilinear-Shaders

**Comment by stenzek** [comment]
Duplicate of https://github.com/stenzek/duckstation/issues/1133
